topologies/proto/binding.proto

```diff
@@ -104,6 +107,9 @@ message Options {
 
   // The session_id for ATE REST API session id
   int32 session_id = 6;
+
+  // gRPC request timeout (second)
+  int32 timeout = 7;
```

Comment:
It looks like `grpc.WithTimeout`  deprecated.
https://pkg.go.dev/google.golang.org/grpc#WithTimeout
So, I have fixed it using
```go
ctx, cancelFunc := context.WithTimeout(context.Background(), time.Duration(d.Timeout)*time.Second)
```
Please let me know if you are ok with it. Otherwise will use `grpc.WithTimeout`

Reply:
Okay, I confused the dial timeout and per-request timeout. grpc.WithTimeout was once the dial timeout, so it's deprecated by grpc.DialContext. But that means the timeout is only observed when using gosnappi.

There is a way to add per-call timeout using go-grpc-middleware.

https://github.com/grpc-ecosystem/go-grpc-middleware/blob/99612e884e71d96a3c8dd30ad4d8eccacfa91798/retry/options.go#L101

The usage around grpc.DialContext would be like this in topologies/binding/options.go dialGRPC():

```
	var retryOpts []grpc_retry.CallOption
	if d.Timeout != 0 {
		retryOpts = append(retryOpts, grpc_retry.WithPerRetryTimeout(d.Timeout * time.Second))
	}
	if len(retryOpts) > 0 {
		opts = append(opts, grpc.WithChainUnaryInterceptor(grpc_retry.UnaryClientInterceptor(retryOpts...)))
	}
	return grpc.DialContext(ctx, d.Target, opts...)
```

Let me know if this works.

Reply:
I tried your suggestion in a scenario where gRPC server is not reachable. And it seems it waiting till the default test timeout (10min) and finally throw this error:
```
panic: test timed out after 10m0s
``` 

According to the description of `grpc_retry.WithPerRetryTimeout` 
```
// The context.Deadline of the call takes precedence and sets the maximum time the whole invocation
// will take, but WithPerRetryTimeout can be used to limit the RPC time per each call.
//
// For example, with context.Deadline = now + 10s, and WithPerRetryTimeout(3 * time.Seconds), each
// of the retry calls (including the initial one) will have a deadline of now + 3s.
```
So, it require `context.WithTimeout` to work properly. here is an example for it https://github.com/grpc-ecosystem/go-grpc-middleware/blob/99612e884e71d96a3c8dd30ad4d8eccacfa91798/retry/examples_test.go#L85

Modified code snippet 
```go
	if d.Timeout != 0 {
		ctx, cancelFunc := context.WithTimeout(context.Background(), time.Duration(d.Timeout)*time.Second)
		defer cancelFunc()
		var retryOpts []grpc_retry.CallOption
		retryOpts = append(retryOpts,
			grpc_retry.WithMax(2),
			grpc_retry.WithPerRetryTimeout(20*time.Second))
		opts = append(opts,
			grpc.WithStreamInterceptor(grpc_retry.StreamClientInterceptor(retryOpts...)),
			grpc.WithUnaryInterceptor(grpc_retry.UnaryClientInterceptor(retryOpts...)),
		)
		return grpc.DialContext(ctx, d.Target, opts...)
	} else {
		return grpc.DialContext(ctx, d.Target, opts...)
	}
```
So, do we require this `WithPerRetryTimeout` or can we go with previous approach `context.WithTimeout`?

Reply:
> I tried your suggestion in a scenario where gRPC server is not reachable. And it seems it waiting till the default test timeout (10min) and finally throw this error:

Is 10 minutes what you set for d.Timeout? Sounds like it's working as intended. If context has a sooner timeout, it will be used. Otherwise the per-request timeout is capped to d.Timeout.

I'm not sure about your modified code snippet. It's hard coding a WithPerRetryTimeout to 20 seconds and doing 2 retries. If we want retries, we should add it as a field in the binding proto.

Reply:
> Is 10 minutes what you set for d.Timeout? Sounds like it's working as intended. If context has a sooner timeout, it will be used. Otherwise the per-request timeout is capped to d.Timeout.

I am talking bout Go’s package testing has a default timeout of 10 minutes (https://terratest.gruntwork.io/docs/testing-best-practices/timeouts-and-logging/)

> I'm not sure about your modified code snippet. It's hard coding a WithPerRetryTimeout to 20 seconds and doing 2 retries. If we want retries, we should add it as a field in the binding proto.

sorry for the typo. I have modified the code snippet. But, I feel only `WithPerRetryTimeout` will not fulfill our use case. We have to specify `context.WithTimeout`.

Reply:
I think we forgot to mention in the documentation that featureprofiles tests need much longer timeout, so our tests should run with maybe `go test -timeout 1h` or even 2h.

Test timeouts are not done through context.WithTimeout. The context timeout mechanism does not panic, whereas the test timeout panics by forcefully killing the goroutine that runs the test function. Using context.Timeout will not prevent test timeout from panicking. These are two completely different mechanisms, so for the purpose of this discussion, we can ignore test timeouts.

Please implement the WithPerRetryTimeout as I requested in https://github.com/openconfig/featureprofiles/pull/720#discussion_r1024608479

Reply:
According to our discussion I have modified the code.